bgpd: Not advertised to any peer in peer-group

soumyar-roy opened the pull request:
Issue:
When we have bgp peer-group, and prefix are advertised to the peers in it correctly, it still shows Not advertised to any peer.

leaf-1# sh bgp ipv6 unicast 2051:51:1:168::/64

BGP routing table entry for 2051:51:1:168::/64, version 69

Paths: (1 available, best #1, table default)

  Not advertised to any peer<<<<<<<<<Wrong state.

  Local

    ::(leaf-1) from :: (6.0.0.17)

      Origin incomplete, metric 0, weight 32768, valid, sourced, bestpath-from-AS Local, best (First path received)

      Last update: Tue Apr  1 18:19:09 2025


Fix:
1)Changed advertised statement to "Advertised to peers:" which is applicable for peer in peer group or non-peer-group
2)Added peer group information in show bgp  json output
3)added testcase
BGP routing table entry for 192.168.251.0/30, version 1
Paths: (1 available, best #1, table default)
  Advertised to peers:
  192.168.255.3 192.168.251.2 192.168.252.2 r1-eth0
  Local
    0.0.0.0 from 0.0.0.0 (192.168.255.1)
      Origin incomplete, metric 0, weight 32768, valid, sourced, best (First path received)
      Last update: Tue Apr 29 17:28:02 2025


BGP advertisement output json:
{
  "prefix":"192.168.251.0/30",
  "version":1,
  "advertisedTo":{
    "192.168.255.3":{
      "hostname":"r3",
      "peerGroup":"PG"
    },
    "192.168.251.2":{
      "hostname":"r2",
      "peerGroup":"PG2"
    },
    "192.168.252.2":{
      "hostname":"r4",
      "peerGroup":"PG2"
    },
    "r1-eth0":{
      "hostname":"r2"
    }
  },
  "paths":[
    {
      "aspath":{
        "string":"Local",
        "segments":[
        ],
        "length":0
      },
      "origin":"incomplete",
      "metric":0,
      "weight":32768,
      "valid":true,
      "version":1,
      "sourced":true,
      "bestpath":{
        "overall":true,
        "selectionReason":"First path received"
      },
      "lastUpdate":{
        "epoch":1744926043,
        "string":"Thu Apr 17 21:40:43 2025"
      },
      "fibInstalled":true,
      "nexthops":[
        {
          "ip":"0.0.0.0",
          "hostname":"r1",
          "afi":"ipv4",
          "metric":0,
          "accessible":true,
          "used":true
        }
      ],
      "peer":{
        "peerId":"0.0.0.0",
        "routerId":"192.168.255.1"
      }
    }
  ]
}

ton31337 commented on bgpd/bgp_route.c:
```diff
@@ -13017,7 +13066,10 @@ void route_vty_out_detail_header(struct vty *vty, struct bgp *bgp,
+	struct hash *peer_groups = NULL;
+	struct process_peer_group_info peer_group_info;
 
+	memset(&peer_group_info, 0, sizeof(peer_group_info));
```
Let's do it this way:

```c
struct process_peer_group_info peer_group_info = {};
```

soumyar-roy replied:
Done

soumyar-roy commented on bgpd/bgp_route.c:
```diff
@@ -12983,6 +12990,48 @@ static void bgp_show_all_instances_routes_vty(struct vty *vty, afi_t afi,
+	json_object *json = peer_group_info->json;
+	struct peer *gpeer = NULL;
+	struct listnode *node;
+	int first = 1;
```
done

soumyar-roy commented on bgpd/bgp_route.c:
```diff
@@ -166,6 +166,13 @@ DEFINE_HOOK(bgp_process,
 	     struct peer *peer, bool withdraw),
 	    (bgp, afi, safi, bn, peer, withdraw));
 
+/* Struct to hold infomation for peer groups processing*/
```
Fixed

soumyar-roy commented:
> Please create a topotest covering this functionality.

Added.

mjstapp commented on bgpd/bgp_route.c:
```diff
@@ -12979,6 +12986,49 @@ static void bgp_show_all_instances_routes_vty(struct vty *vty, afi_t afi,
+{
+	const struct peer_group *g1 = p1;
+	const struct peer_group *g2 = p2;
+	return strcmp(g1->name, g2->name) == 0;
```
needs a blank line

mjstapp commented on bgpd/bgp_route.c:
```diff
@@ -12979,6 +12986,49 @@ static void bgp_show_all_instances_routes_vty(struct vty *vty, afi_t afi,
+static unsigned int peer_group_hash_key(const void *p)
+{
+	const struct peer_group *group = p;
+	return string_hash_make(group->name);
```
needs a blank line

mjstapp commented on bgpd/bgp_route.c:
```diff
@@ -11248,7 +11255,7 @@ void route_vty_out_detail(struct vty *vty, struct bgp *bgp, struct bgp_dest *bn,
 	json_object *json_string = NULL;
 	json_object *json_adv_to = NULL;
-	int first = 0;
+	bool first = false;
```
so ... you've change the type here, but not the way the variable is set later on? are there other examples of that?

soumyar-roy replied:
There was another place inside this function, I have changed that too now. Thanks for finding.

mjstapp commented on bgpd/bgp_route.c:
```diff
@@ -13181,8 +13233,19 @@ void route_vty_out_detail_header(struct vty *vty, struct bgp *bgp,
+				if (bgp_adj_out_lookup(peer, dest, 0)) {
+					/* Create a hash table to track peer groups */
+					if (!peer_groups)
+						peer_groups = hash_create(peer_group_hash_key,
```
typesafe collections, please

soumyar-roy replied:
Changed ti DECLARE_HASH type

mjstapp commented on bgpd/bgp_route.c:
```diff
@@ -13181,8 +13233,19 @@ void route_vty_out_detail_header(struct vty *vty, struct bgp *bgp,
+				if (bgp_adj_out_lookup(peer, dest, 0)) {
+					/* Create a hash table to track peer groups */
+					if (!peer_groups)
+						peer_groups = hash_create(peer_group_hash_key,
```
why a hash here - it sort of looks as if the collection is just used to iterate through later on?
and I guess ... why not just emit the group info in-line here, as individual peers are done? the PR has examples of a problem, but not much about why this particular change should be done.

soumyar-roy replied:
the aim is to show 1) create one header for all non group peers the show the non peer group peers 2) later show each peer group and all peers under a group. So I don't understand what does this mean "why not just emit the group info in-line here, as individual peers are done?". I dont want to mix the peers with some peer from a group.

mjstapp replied:
hmm. ok, well, maybe it would be easier just to emit each peer - maybe the old/existing pattern that uses that header is not really correct? your comment reads as if you think that retaining that header line is the most important thing - but maybe it's ... not?

soumyar-roy replied:
It was wrongly "showing Advertised to non peer-group peers:" where the peers were actually part of group>>That tried to be fixed by 3fb72a03c3 >>But it caused to wrongly showing "Not advertised to any peer", when actually it was sending to peers part of group. So header is important here.

mjstapp replied:
so ... I think the json output, for example, just includes ... all the peers, and that seems most correct. perhaps that's what should be done for the vty output too: just include the peers, acknowledge that there will be group and non-group peers in the list. maybe the group name could be included with a group peer, in case that's interesting?

soumyar-roy replied:
1) One of things considered, when there are huge number of V6 peers , whether to dump them all. This design leaves that option open, either we can show all peers or certain number of peers or just show the peer group name, but dont show the peers explicitly. We can adjust the code accordingly  2) this will also help to debug when some policy applied to some group . Sample CISCO XR code given for ref >>show bgp ipv4 labeled-unicast 10.100.1.5
BGP routing table entry for 10.100.1.5/32
Versions:
  Process           bRIB/RIB  SendTblVer
  Speaker                 76          76
    Local Label: 24007
Last Modified: Apr  8 09:37:57.509 for 00:04:05
Paths: (1 available, best #1)
  Advertised to update-groups (with more than one peer):
    0.3
  Advertised to peers (in unique update groups):
    10.100.1.1      10.100.1.2     
  Path #1: Received by speaker 0
  Advertised to update-groups (with more than one peer):
    0.3
  Advertised to peers (in unique update groups):
    10.100.1.1      10.100.1.2     
  2
    10.100.1.4 from 10.100.1.4 (10.100.1.4)
      Received Label 24000
      Origin incomplete, metric 2, localpref 100, valid, external, best, group-best
      Received Path ID 0, Local Path ID 1, version 76
      Origin-AS validity: not-found

Considering I will change the code to use type safe container /handle other comments, Can you please clarify, what are the issues we can face/your concern with the code with hash, I have currently.

soumyar-roy replied:
Suppose we have  mix of non group and non group peers, p1(non group), p2(group), p3(group), p4(non group), p5(non group), whats your proposal to handle that?. "Advertised to non peer-group peers:" string doesn't work.

soumyar-roy replied:
As discussed, reverted all changes. now 1)Changed advertised statement to "Advertised to peers:" which is applicable for peer in peer group or non-peer-group
2)Added peer group information is show bgp vty and json output
3)added testcase

soumyar-roy commented:
Hi All, Is there anything more to take care for this,  from my side?. We need to use this fix soon. 
As discussed, reverted all changes. now 1)Changed advertised statement to "Advertised to peers:" which is applicable for peer in peer group or non-peer-group
2)Added peer group information is show bgp vty and json output
3)added testcase

Thanks.

ton31337 commented on tests/topotests/bgp_peer_group/test_bgp_peer-group.py:
```diff
@@ -179,6 +180,141 @@ def _bgp_peer_group_remoteas_add():
+
+    r1 = tgen.gears["r1"]
+    logger.info("Change a peer r1-eth0 to be non peer group and PG2 to have two peers")
+    r1.cmd(
```
Please use directly `r1.vtysh_cmd()`.

ton31337 commented on tests/topotests/bgp_peer_group/test_bgp_peer-group.py:
```diff
@@ -179,6 +180,141 @@ def _bgp_peer_group_remoteas_add():
+    )
+
+    # Wait for BGP to converge
+    sleep(5)
```
No sleep() please in the topotests.

soumyar-roy replied:
Changed.

ton31337 commented on tests/topotests/bgp_peer_group/test_bgp_peer-group.py:
```diff
@@ -179,6 +180,141 @@ def _bgp_peer_group_remoteas_add():
+        'vtysh -c "show bgp ipv4 unicast 192.168.251.0/30 json"'
+    ).rstrip()
+
+    import re
```
Move imports to the top.

soumyar-roy replied:
Done

soumyar-roy commented:
> ```
> Advertised to peers:
>   192.168.255.3[Peer Group:PG] 192.168.251.2[Peer Group:PG2] 192.168.252.2[Peer Group:PG2] r1-eth0
> ```
> 
> This is going to be a mess if we have lots of neighbors in a peer-group. IMO, we should to reverse...
> 
> ```
> Advertised to peers:
>   PG(192.168.251.1) PG2(92.168.251.2, 192.168.252.2) r1-eth0
> ```

I agree, so when I started the PR, the output was similar to this, especially considering scaled neighbors>>
BGP routing table entry for 192.168.251.0/30, version 1
Paths: (1 available, best #1, table default)
  Advertised to non peer-group peers:
  r1-eth0

  Advertised to peer-group PG peers:
  192.168.255.3

  Advertised to peer-group PG2 peers:
  192.168.252.2 192.168.251.2
>>Later I was suggested to current way, personally I dont like this way though, i.e. 1) 192.168.255.3[Peer Group:PG] 192.168.251.2[Peer Group:PG2] 192.168.252.2[Peer Group:PG2] r1-eth0, I still prefer show all peers per peer group, similar to the way you mentioned 2) PG(192.168.251.1) PG2(92.168.251.2, 192.168.252.2) r1-eth0. Can we have an agreement on this?.

ton31337 commented:
I vote for your initial output:

```
Advertised to non peer-group peers:
r1-eth0

Advertised to peer-group PG peers:
192.168.255.3

Advertised to peer-group PG2 peers:
192.168.252.2 192.168.251.2
```

or my suggested:

```
Advertised to peers:
  PG(192.168.251.1) PG2(92.168.251.2, 192.168.252.2) r1-eth0
```

@mjstapp what do you think?

soumyar-roy commented:
Addressed all comments. As discussed today, removed changes to show peer-group info from vty output, but kept only in json output in this PR, a separate PR can be raised to handle peer group info in vty output, if needed later. Fixed the original bug.

ton31337 commented on tests/topotests/bgp_peer_group/test_bgp_peer-group.py:
```diff
@@ -17,6 +17,9 @@
 import json
 import pytest
 import functools
+from time import sleep
```
Please drop sleep()s.

soumyar-roy replied:
Done.

mjstapp commented on tests/topotests/bgp_peer_group/test_bgp_peer-group.py:
```diff
@@ -179,6 +181,117 @@ def _bgp_peer_group_remoteas_add():
+    """
+    )
+
+    logger.info(f"test_bgp_peer_group_with_non_peer_group_peer passed")
```
so ... the change to json is not tested at all?

soumyar-roy replied:
Yes, it is tested in the beginning >>return topotest.json_cmp(output, expected), I have also added json output in PR description.

ton31337 replied:
Can we drop this `logger.info` at all?

soumyar-roy replied:
Removed logger.info(f"test_bgp_peer_group_with_non_peer_group_peer passed")

mjstapp commented on bgpd/bgp_route.c:
```diff
@@ -11117,9 +11117,8 @@ static void flap_route_vty_out(struct vty *vty, const struct prefix *p,
-					int *first, const char *header,
-					json_object *json_adv_to)
+static void route_vty_out_advertised_to(struct vty *vty, struct peer *peer, int *first,
+					const char *header, json_object *json_adv_to)
```
white-space change?

soumyar-roy replied:
Fixed

mjstapp reviewed (approved):
the bgp json change is fine, thanks

soumyar-roy commented on tests/topotests/bgp_peer_group/test_bgp_peer-group.py:
```diff
@@ -179,6 +181,117 @@ def _bgp_peer_group_remoteas_add():
+        output_str = r1.vtysh_cmd("show bgp ipv4 unicast 192.168.251.0/30 json")
+        try:
+            output = json.loads(output_str)
+        except json.JSONDecodeError:
```
Removed.

soumyar-roy commented:
Hi All, if no other comments, can you please help to merge it?. Thanks.
